### PR TITLE
RSDK-4521 RSDK-4729 Improve existing module examples

### PR DESCRIPTION
**Pull request by benjirewis**
[RSDK-4521](https://viam.atlassian.net/browse/RSDK-4521)
[RSDK-4729](https://viam.atlassian.net/browse/RSDK-4729)

Reorganizes example modules into subdirectories. Renames example module to simple module. Modifies simple module to pull a `to_print` value from attributes instead of just increasing ID number. Fixes (hopefully) simple module's signal handling and threading.

[RSDK-4521]: https://viam.atlassian.net/browse/RSDK-4521?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RSDK-4729]: https://viam.atlassian.net/browse/RSDK-4729?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/viam/examples/modules/CMakeLists.txt [hunk 1]
```diff
@@ -12,58 +12,5 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-add_executable(example_module
```
Comment (benjirewis) [inline src/viam/examples/modules/CMakeLists.txt:15]: This cmake logic has been split between `src/viam/examples/modules/tflite/CMakeLists.txt` and `src/viam/examples/modules/simple/CMakeLists.txt`.

File: src/viam/examples/modules/simple/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+For more information, see the [documentation](https://docs.viam.com/program/extend/modular-resources/). For a more complex example, take a look at the [complex module example](https://github.com/viamrobotics/viam-cpp-sdk/tree/main/src/viam/examples/modules/complex), which contains multiple new APIs and custom resource models.
+
+## Project structure
+The `main.cpp` file contains the definition of a new generic model and code to register it. It also has the optional validator function and implements reconfigure. The validator function is defined upon resource registration, and the reconfigure method is implemented on the resource class.
+
+The validator function can throw errors that are triggered due to errors in the configuration. It also returns a vector of strings representing the implicit dependencies of the resource. Note that printers have no implicit dependencies; see the [complex module example](https://github.com/viamrobotics/viam-cpp-sdk/tree/main/src/viam/examples/modules/complex) for examples of modular resources with implicit dependencies.
```
Comment (benjirewis) [inline src/viam/examples/modules/simple/README.md:14]: The complex module example does not exist yet but will hopefully be available through this URL after https://viam.atlassian.net/browse/RSDK-4522.

File: src/viam/examples/modules/simple/main.cpp [hunk 3]
```diff
@@ -24,118 +23,126 @@
...
-    // shuts down your module. pthread_sigmask should be called near the start
-    // of main so that later threads inherit the mask.
-    sigset_t sigset;
-    sigemptyset(&sigset);
-    sigaddset(&sigset, SIGINT);
-    sigaddset(&sigset, SIGTERM);
-    pthread_sigmask(SIG_BLOCK, &sigset, NULL);
+    // C++ modules must handle SIGINT and SIGTERM. You can use the SignalManager
+    // class and its wait method to handle the correct signals.
+    SignalManager signals;
```
Comment (benjirewis) [inline src/viam/examples/modules/simple/main.cpp:110]: I copied the signal handling logic from @abe-winter's wifi example CPP module. This should offset some of the burden of boilerplate signal handling from the user.

File: src/viam/examples/modules/simple/main.cpp [hunk 4]
```diff
@@ -24,118 +23,126 @@
...
-    auto my_mod = std::make_shared<ModuleService_>(argv[1]);
+    auto my_mod = std::make_shared<ModuleService_>(socket_addr);
     auto server = std::make_shared<Server>();
 
     my_mod->add_model_from_registry(server, generic, m);
     my_mod->start(server);
+    BOOST_LOG_TRIVIAL(info) << "Module serving model " << m.to_string() << ", listening on "
+                            << socket_addr;
 
-    std::thread server_thread([&server, &sigset]() {
```
Comment (benjirewis) [inline src/viam/examples/modules/simple/main.cpp:130]: I now start the server synchronously; @abe-winter's example module does this, and this simple module seems to run/shutdown just fine with the RDK. This seems to have gotten rid of the issue in https://viam.atlassian.net/browse/RSDK-4729.

File: src/viam/examples/modules/simple/main.cpp [hunk 5]
```diff
@@ -24,118 +23,126 @@
 using viam::component::generic::v1::GenericService;
 using namespace viam::sdk;
 
-class MyModule : public GenericService::Service, public Component {
+// Printer is a modular resource that can print a to_print value to STDOUT when
+// a DoCommand request is received or when reconfiguring. The to_print value
+// must be provided as an attribute in the config.
+class Printer : public GenericService::Service, public Component {
```
Comment (stuqdog) [inline src/viam/examples/modules/simple/main.cpp:29]: from speaking with @abe-winter, I think this should properly inherit from `Generic`, not `GenericService::Service`. Seems like a bit more of a refactor though.

File: src/viam/examples/modules/tflite/README.md
Comment (stuqdog) [inline src/viam/examples/modules/tflite/README.md:1]: I think adding a comprehensive README is good but maybe not super pressing given how comprehensive the comments are in the tflite example itself. Perhaps worth adding a note to look at the comments in the code for understanding?

File: src/viam/sdk/module/service.hpp [hunk 6]
```diff
@@ -56,5 +58,22 @@ class ModuleService_ : public viam::module::v1::ModuleService::Service {
...
+ * C++ modules must handle SIGINT and SIGTERM. Make sure to create a sigset
+ * for SIGINT and SIGTERM that can be later awaited in a thread that cleanly
+ * shuts down your module. pthread_sigmask should be called near the start of
+ * main so that later threads inherit the mask. SignalManager can handle the
+ * boilerplate of this functionality.
+ */
+class SignalManager {
+   public:
+    SignalManager();
+    /** wait on sigset */
```
Comment (stuqdog) [inline src/viam/sdk/module/service.hpp:71]: (very much a nit) just for consistency, can we indicate comments with `//` rather than `/**`?

Comment (benjirewis) [reply src/viam/sdk/module/service.hpp:71]: Yep!

File: src/viam/examples/modules/simple/main.cpp (hunk 5 above)
Comment (benjirewis) [reply src/viam/examples/modules/simple/main.cpp:29]: Good point. Switched it to inherit from only `Generic`. Required some refactoring, but things seem simpler now.

File: src/viam/examples/modules/tflite/README.md
Comment (benjirewis) [reply src/viam/examples/modules/tflite/README.md:1]: Yeah good point. Just elaborated a bit more here and referred users to inline comments. Should I just close RSDK-4833 as "won't do" for now?

Comment (stuqdog) [reply src/viam/examples/modules/tflite/README.md:1]: Yeah I think that sounds good, thanks!

**Review by stuqdog — APPROVED** [review]
lgtm, thanks for putting this together!

File: src/viam/examples/modules/simple/README.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+			"name": "MyModule",
+			"executable_path": "/home/viam-cpp-sdk/build/install/bin/example_module"
+		}
+	],
+	"components": [
+		{
+			"namespace": "rdk",
+			"type": "generic",
+			"name": "printer1",
+			"model": "acme:demo:printer",
```
Comment (cheukt) [inline src/viam/examples/modules/simple/README.md:39]: the model should be viam:generic:printer as of `https://viam.atlassian.net/browse/RSDK-4648`

**Review by cheukt — APPROVED** [review]
requesting a small change otherwise looks great!

**Pull request merged**